core/extension/gdextension_interface.cpp

```diff
@@ -1041,7 +1041,43 @@ static void gdextension_ref_set_object(GDExtensionRefPtr p_ref, GDExtensionObjec
 	ref->reference_ptr(o);
 }
 
+#ifndef DISABLE_DEPRECATED
 static GDExtensionScriptInstancePtr gdextension_script_instance_create(const GDExtensionScriptInstanceInfo *p_info, GDExtensionScriptInstanceDataPtr p_instance_data) {
+	const GDExtensionScriptInstanceInfo2 info_2 = {
+		p_info->set_func,
+		p_info->get_func,
+		p_info->get_property_list_func,
+		p_info->free_property_list_func,
+		p_info->property_can_revert_func,
+		p_info->property_get_revert_func,
+		p_info->get_owner_func,
+		p_info->get_property_state_func,
+		p_info->get_method_list_func,
+		p_info->free_method_list_func,
+		p_info->get_property_type_func,
+		p_info->has_method_func,
+		p_info->call_func,
+		nullptr, // notification_func.
+		p_info->to_string_func,
+		p_info->refcount_incremented_func,
+		p_info->refcount_decremented_func,
+		p_info->get_script_func,
+		p_info->is_placeholder_func,
+		p_info->set_fallback_func,
+		p_info->get_fallback_func,
+		p_info->get_language_func,
+		p_info->free_func,
+	};
+
+	ScriptInstanceExtension *script_instance_extension = memnew(ScriptInstanceExtension);
+	script_instance_extension->instance = p_instance_data;
+	script_instance_extension->native_info = &info_2;
```

Comment:
This is taking the address of a temporary, which is causing crashes when the functions are called later.

Reply:
Ack, right because unlike `classdb_register_extension_class` we're not copying the data into a new internal struct, we're directly referring to the struct provided by the GDExtension. We should probably not do that - it'd probably be better to copy that data into an instance of the struct on the Godot side...

Reply:
Copying the data might be better in general. This is per instance tho, so either we duplicate the memory for each instance, or somehow create a mapping...

Another solution for this specific case is to add another class with a virtual override Godot side:

- `ScriptInstance`
- `ScriptInstanceExtensionDeprecated` (add this)
- `ScriptInstanceExtension`

Then have `gdextension_script_instance_create[2]` create the correct class.

Reply:
Here's a proposed fix that copies the data: https://github.com/godotengine/godot/pull/81205

> This is per instance tho, so either we duplicate the memory for each instance, or somehow create a mapping...

Hm, yeah, I suppose this does make each instance take up 23 pointers more memory... This is a trickier problem than I thought at first!

Reply:
Sorry about that. Looks like I don't have a good intuition for these kind of problems.

Reply:
> Another solution for this specific case is to add another class with a virtual override Godot side

Yeah, I worry about the scalability of this approach: what happens as we make more changes to this struct? (We already have a change queued up for https://github.com/godotengine/godot/issues/81119)

I'd like to avoid the compatibility code continuing to require maintenance and grow in complexity over time. Copying is at least simple and straight forward. But copying as done in my current PR is always a performance and memory hit, even on the new code path. What if the hit was only on the compatibility path, maybe that's an OK trade-off?

Directly relates this comment on the PR: https://github.com/godotengine/godot/pull/81205#issuecomment-1701686152

Reply:
> Sorry about that. Looks like I don't have a good intuition for these kind of problems.

No worries - several other people (including me) read through your code without noticing it :-)